api/proxy/proxy.go

```diff
@@ -0,0 +1,391 @@
+			p.listenAddrPort = netip.AddrPortFrom(p.listenAddrPort.Addr(), uint16(p.sessionAuthzData.DefaultClientPort))
+		}
+	}
+	p.connectionsLeft.Store(p.sessionAuthzData.ConnectionLimit)
+	p.workerAddr = p.sessionAuthzData.WorkerInfo[0].Address
+
+	tlsConf, err := p.clientTlsConfig(opt...)
+	if err != nil {
+		return nil, fmt.Errorf("error creating TLS configuration: %w", err)
+	}
+	p.createTime = p.sessionAuthzData.CreatedTime
+	p.expiration = p.sessionAuthzData.Expiration
+
+	// We don't _rely_ on client-side timeout verification but this prevents us
+	// seeming to be ready for a connection that will immediately fail when we
+	// try to actually make it
+	p.ctx, p.cancel = context.WithDeadline(ctx, p.expiration)
+
+	transport := cleanhttp.DefaultTransport()
+	transport.DisableKeepAlives = false
+	// This isn't/shouldn't used anyways really because the connection is
+	// hijacked, just setting for completeness
+	transport.IdleConnTimeout = 0
+	transport.DialContext = func(ctx context.Context, network, addr string) (net.Conn, error) {
+		dialer := &tls.Dialer{Config: tlsConf}
+		return dialer.DialContext(ctx, network, addr)
+	}
+	p.transport = transport
+
+	return p, nil
+}
+
+// Start starts the listener for client proxying. It ends, with any errors, when
+// the listener is closed and no connections are left. Cancel the client's proxy
+// to force this to happen early. It is not safe to call Start twice, including
+// once it has exited, and will immediately error in this case; create a new
+// ClientProxy with New().
+//
+// Note: if a custom listener implementation is used and the implementation can
+// return a Temporary error, the listener will not be closed on that condition
+// and no feedback will be given. It is up to the listener implementation to
+// inform the client, if needed, of any status causing a Temporary error to be
+// returned on accept.
+//
+// EXPERIMENTAL: While this API is not expected to change, it is new and
+// feedback from users may necessitate changes.
+func (p *ClientProxy) Start() (retErr error) {
+	if !p.started.CompareAndSwap(false, true) {
+		return errors.New("proxy was already started")
+	}
+
+	defer p.cancel()
+
+	if p.listener.Load() == nil {
+		var err error
+		ln, err := net.ListenTCP("tcp", &net.TCPAddr{
+			IP:   p.listenAddrPort.Addr().AsSlice(),
+			Port: int(p.listenAddrPort.Port()),
+		})
+		if err != nil {
+			return fmt.Errorf("unable to start listening: %w", err)
+		}
+		p.listener.Store(ln)
+	}
+
+	listenerCloseFunc := func() {
+		p.listenerCloseOnce.Do(func() {
+			// Forces the for loop to exit instead of spinning on errors
+			p.cancel()
+			p.connectionsLeft.Store(0)
+			if err := p.listener.Load().(net.Listener).Close(); err != nil && err != net.ErrClosed {
+				retErr = errors.Join(retErr, fmt.Errorf("error closing proxy listener: %w", err))
+			}
+		})
+	}
+
+	// Ensure closing the listener runs on any other return condition
+	defer listenerCloseFunc()
+
+	fin := make(chan error, 10)
```

Comment:
Likely 3 or 4 would be enough. We don't drain the channel until things close, but there are only a few actual places we can put an error on the channel. So this is 10 for safety even though it can only top out at a few.